### DISCUSSION TITLE
Implement a htmx gomponents library

### DISCUSSION THREAD

**Discussion by lobre** [maragudk/gomponents#128] [Ideas] [answered]
Hey,

I love the concepts of this project.

I have just read the article on goingwithgo.com and it seems you mention the fact that you might want to create an "official" htmx library!.

https://goingwithgo.com/2023/02/building-reusable-web-view-components-in-pure-go-markus-wustenberg/

I just wanted to know if this was still on your plate. And if so, I was thinking that this discussion could track the topic ;)

I also see that another project was created some time ago but it seems it is not in active development anymore.

https://github.com/flaticols/gomponents-htmx

**Comment by lobre** [comment]
I also see that this other discussion talks a bit about this.

https://github.com/maragudk/gomponents/discussions/113

**Comment by markuswustenberg** [comment]
Yeah, I’m still open to building it. I don’t know if I’ll have full coverage of all features (the hyperscript parts for example), but maybe less is also fine. 

Any parts in particular you’re interested in?

**Reply by lobre** [reply]
Not a part specifically, but the basics of htmx would already be a good start! To me, if it is an "official" library, people will be more likely to contribute new features if they are missing, compared to a repository that has been created in a one-shot without maintenance for years.

Same but different, do you see any value in building a tailwind library, instead of adding the classes as "strings"?

**Reply by markuswustenberg** [reply]
Yeah, fully agreed. I also like the idea of an "official" library. That doesn't necessarily mean I'm the only maintainer/contributor, of course. 😊

I've set up a repo at https://github.com/maragudk/gomponents-htmx now. No work done yet.

**Reply by markuswustenberg** [reply]
I've considered a tailwind one as well, but I don't feel there's a lot I can contribute, except a lot of constants. And there are _a lot_, and Tailwind is developing really fast.

What opportunities do you see for a gomponents-tailwind library? Maybe there's something I'm missing.

**Reply by lobre** [reply]
This question was just out of curiosity. As you say, I agree this would mostly be constants. And I think that having functions/constants is superior to having strings as the go compiler can help to spot mistakes.

But apart from this, not much at the first sight. And as you say, they have a fast development pace, so might be hard to catch up and stay up-to-date.

**Reply by markuswustenberg** [reply]
Yeah, agreed. I'd have to try it out to see how the API feels, but since there are so many strings that are basically concatenated, I think it would be a bit of a mess with just constants. Could be more interesting perhaps with functions that compose, so e.g. `bg-gray-50` turns into `Bg(Gray(50))` or something. Hmm. I'll think about it. Thank you for the feedback!

**Reply by markuswustenberg** [reply]
Okay, I've added all the core attributes from https://htmx.org/reference/#attributes . I think http headers will go into an `http` subpackage, to distinguish from the attributes. It's actually going to be less work than I expected. 😄

**Reply by lobre** [reply]
Awesome! This was fast 🚀

I can see that attributes such as `hx-put` are considered as "Additional". For a complete list, might be still worth it to include those as well in the future.

https://htmx.org/reference/#attributes-additional

**Reply by lobre** [reply]
But anyway this reference documentation is not that long, I thought there were many more concepts and utilities in htmx, but not that much!

**Reply by markuswustenberg** [reply]
Yeah, I’m going to include everything I think. It’s not a lot. I just want to make sure I get the API right, but after playing with it today, I’m pretty confident about it. Watch the repo for changes. 😊

**Reply by lobre** [reply]
I will totally subscribe to the repo to listen for modifications. Thanks a lot for implementing this!

**Comment by flaticols** [comment]
Hi all, we are using gomponents and htmx at our company. Maybe I can help with that.

**Reply by markuswustenberg** [reply]
Cool, thank you! I've created the repository at https://github.com/maragudk/gomponents-htmx . There's nothing done yet, but I think I'll have time to work on it in the coming weeks.

Just out of curiosity, what company is that? I'm thrilled when someone is using my libraries. 😊

**Reply by flaticols** [reply]
I'll be making some commits this weekend from my library https://github.com/flaticols/gomponents-htmx.

We are a small (~40 people) company from Amsterdam. Right now we use gomponents & htmx for an internal product. We have a small "framework". Everything works smoothly, fast, fantastic.

**Reply by ossy45** [reply]
> Hi all, we are using gomponents and htmx at our company. Maybe I can help with that.

Okay what the name of your company

**Comment by markuswustenberg** [comment, marked as answer]
For people finding this and TLDR: https://www.gomponents.com/plus/#htmx
